### PR TITLE
BaseOpenIdJWTAuthentication should skip iss/aud bad matches

### PR DESCRIPTION
**Pull request by ericbuckley**
because this authentication module could be used more than once in a
project, if the aud and/or iss doesn't match, we should return None and
move on to the next authentication module rather than raise an error

### CODE REVIEW

File: drftoolbox/authentication.py [hunk 1]
```diff
@@ -118,25 +118,38 @@ def get_jwt_value(self, request):
...
+    def decode_handler(self, token):
+        header = jose_jwt.get_unverified_header(token)
+        claims = jose_jwt.get_unverified_claims(token)
+        key = self.get_public_key(claims.get('iss'), kid=header.get('kid'))
+        issuers = self.acceptable_issuers()
+        audience = self.acceptable_audience(claims)
+        if not issuers:
+            raise NotImplementedError('at least one issuer must be defined')
+        if claims.get('iss') not in issuers:
+            raise TypeError('invalid issuer')
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply drftoolbox/authentication.py]: We could do that, I don't think there are other possibilities, but I didn't really audit the `get_public_key` method either, so its possible

File: drftoolbox/authentication.py [hunk 2]
```diff
@@ -56,12 +62,13 @@ def openid_configuration_to_jwks_uri(url, timeout=None):
     value = cache.get(key)
     if value is None:
         LOGGER.debug('loading openid configuration')
-        resp = urllib.request.urlopen(url)
+        # TODO: need to catch HTTPError
```
Comment (smcoll) [inline drftoolbox/authentication.py]: you can drop this TODO

File: drftoolbox/authentication.py [hunk 3]
```diff
@@ -118,25 +118,38 @@ def get_jwt_value(self, request):
...
                 required_keys=self.jwks_required_keys, timeout=self.timeout)
         if key is None:
             msg = _('Invalid issuer JWKS URI')
             raise exceptions.AuthenticationFailed(msg)
+        return key
+
+    def decode_handler(self, token):
+        header = jose_jwt.get_unverified_header(token)
+        claims = jose_jwt.get_unverified_claims(token)
+        key = self.get_public_key(claims.get('iss'), kid=header.get('kid'))
```
[the remark this answers is not in the document]
Comment (smcoll) [reply drftoolbox/authentication.py]: And what about moving the `key` assignment to right before `return jose_jwt.decode()`?  We only need to obtain the key (which might involve network IO) if the issuers and audience are valid and match.

[the remark this answers is not in the document]
Comment (ericbuckley) [reply drftoolbox/authentication.py]: good point

File: drftoolbox/authentication.py [hunk 4]
```diff
@@ -155,14 +153,12 @@ def decode_handler(self, token):
...
-            raise JWTMismatchClaimException('invalid issuer')
-        if audience and claims.get('aud') != audience:
-            raise JWTMismatchClaimException('invalid audience')
         key = self.get_public_key(claims.get('iss'), kid=header.get('kid'))
+        if key is None:
+            raise jose_exceptions.JWTClaimsError('missing public key')
         return jose_jwt.decode(
             token,
-            key,
+            key or '',
```
[the remark this answers is not in the document]
Comment (ericbuckley) [reply drftoolbox/authentication.py]: good eye, yeah that was a testing mistake.  `''` will never be evaluated, so I'll remove

**Pull request merged**